New Entry: Big-O notation (#823)
Codecademy/docs: content/general/concepts/algorithm/BigO.md

---

> Title: 'Big-O Notation'

Reviewer: In order to properly nest under "algorithm" this file needs to be relocated to the following path:
content/general/concepts/algorithm/terms/BigO/BigO.md. This will allow Docs to properly put a link in the "algorithm" entry.

Reviewer 2: @SSwiniarski I actually fixed this myself 😄 (noted [here](https://github.com/Codecademy/docs/pull/823#issuecomment-1166311367))

> Description: 'An algorithm is a formal process used to solve a problem. To compare two algorithms, it is necessary to develop tools to evaluate algorithmic efficiency. This comparison is very important since while many algorithms might function in the same way, they may also consume a differing amount of resources. It is often best practice to limit the footprint of the algorithms that are developed and to do that, Big-O notation is to be used. '

Reviewer: ```suggestion
Description: 'Big-O notation is a form of measuring the complexity of an algorithm. '
```

Reviewer: The Description of an entry should A) Immediately reference the term in question and B) Be relatively short and pithy

Subjects:
  - 'Computer Science'
  - 'Web Development'
Tags:
  - 'Algorithms'
CatalogContent:
  - 'paths/computer-science'
  - 'paths/front-end-engineer-career-path'
---

> Big-O notation is a form of measuring the complexity of an algorithm. This can take the form of a time complexity analysis or a space complexity analysis. At times, an algorithm can be refactored or reworked to achieve a more efficient time and space complexity. To describe the time or space complexity, two different notations are used: big omega notation and big-O. Big omega describes the best running case scenario while big-O describes the worst case scenario. 

Reviewer: ```suggestion
**Big-O notation** is a form of measuring the complexity of an algorithm. This can take the form of a time complexity analysis or a space complexity analysis. At times, an algorithm can be refactored or reworked to achieve a more efficient time and space complexity. To describe the time or space complexity, two different notations are used: big omega notation and big-O. Big omega describes the best running case scenario while big-O describes the worst case scenario. 
```

## Common Runtimes

Here are some common runtimes:

> ![graft of the different kinds of big O notation](https://raw.githubusercontent.com/Codecademy/docs/main/media/Big-O_Graph.png)

Reviewer: ```suggestion
![graph of the different kinds of big O notation](https://raw.githubusercontent.com/Codecademy/docs/main/media/Big-O_Graph.png)
```

Reviewer: The media here is lifted from another website. (one example [here](https://levelup.gitconnected.com/big-o-time-complexity-what-it-is-and-why-it-matters-for-your-code-6c08dd97ad59)). It should be replaced with an original graphic.

> <img title="a title" alt="Alt text" src="https://raw.githubusercontent.com/Codecademy/docs/main/media/Big-O_Graph.png" width="500"/>

Reviewer 3: ```suggestion
![graft of the different kinds of big O notation](https://raw.githubusercontent.com/Codecademy/docs/main/media/Big-O_Graph.png)
```

We like using MD for 99% of things including images.

> - Polynomial time: If the time is a power of the input size.

Reviewer: This list would benefit from having examples of the actual notation (O(n) etc) that could tie back to the graph.

  - Bubble sort has quadratic time complexity.
- Linear time: If the time is proportional to the input size.
  - The traverse of a list.
- Logarithmic time: If the time is a logarithmic function of the input size.
  - Binary search algorithm.
- Constant time: If the time needed by the algorithm is the same, regardless of the input size.
  - An access to an array element.

## Simplifying Big-O Expressions

> When determining an expression that charecterizes the time or space complexity of an algorithm, you may find that your expression contains multiple terms. For example: O(n) + O(ln(n)). In this situation, it is best practice to drop all of the lower order terms so the big-O notation would simplify into just O(n).

Reviewer: ```suggestion
When determining an expression that characterizes the time or space complexity of an algorithm, an expression may contain multiple terms. For example: O(n) + O(ln(n)). In this situation, it is best practice to drop all of the lower order terms so the big-O notation would simplify into just O(n).
```

Reviewer: We prefer not to use "you" and "we" in Docs entries.
